File: plugins/scrape/src/main/java/no/lillehaug/landingsopplysninger/scrape/ScrapingJob.java
package no.lillehaug.landingsopplysninger.scrape;

import com.codahale.metrics.MetricRegistry;
import com.codahale.metrics.Timer;
import com.codahale.metrics.health.HealthCheck;
import com.codahale.metrics.health.HealthCheckRegistry;
import no.lillehaug.landingsopplysninger.api.LandingsopplysningerRepository;
import no.lillehaug.landingsopplysninger.api.Leveringslinje;
import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
import org.apache.http.impl.nio.client.HttpAsyncClients;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.time.LocalDate;
import java.util.List;

public class ScrapingJob {

    private static final Logger logger = LoggerFactory.getLogger(ScrapingJob.class);

    private final Timer timer = new Timer();

    private final Scraper scraper;
    private final LandingsopplysningerRepository repository;
    private final List<String> registrations;
    private final HealthCheckRegistry healthCheckRegistry;
    private final MetricRegistry metricRegistry;

    private boolean previousRunSuccess = true;

    ScrapingJob(Scraper scraper, LandingsopplysningerRepository repository, List<String> registrations, HealthCheckRegistry healthCheckRegistry, MetricRegistry metricRegistry) {
        this.scraper = scraper;
        this.repository = repository;
        this.registrations = registrations;
        this.healthCheckRegistry = healthCheckRegistry;
        this.metricRegistry = metricRegistry;

        metricRegistry.register("ScrapeTimerMetrics", timer);
        healthCheckRegistry.register("ScrapingJobHealthCheck", new HealthCheck() {
            @Override
            protected Result check() throws Exception {
                if(previousRunSuccess) {
                    return Result.healthy();
                } else {
                    return Result.unhealthy("ScrapingJob failed");
                }
            }
        });
    }

    void scrapeForRegistrations() {
        try {
            timer.time(() -> {
                try(CloseableHttpAsyncClient httpclient = HttpAsyncClients.createDefault()) {
                    logger.info("Running scrapeForRegistrations for {}", registrations);
                    for (String registration: registrations) {
                        scrapeForRegistration(registration, httpclient);
                    }

                } catch(Throwable e) {
                    previousRunSuccess = false;
                    logger.error("Error scrapeForRegistrations", e);
                }
                logger.info("Done running scrapeForRegistrations for {}", registrations);
                return null;
            });
        } catch (Exception e) {
            previousRunSuccess = false;
            logger.error("Error scrapeForRegistrations", e);
        }
    }

    private void scrapeForRegistration(String registration, CloseableHttpAsyncClient httpclient) {
        logger.info("Running scrapeForRegistrations for {}", registration);
        Mono<LocalDate> forrigeLanding = repository.forrigeLandingFor(registration);
        Flux<Leveringslinje> landinger = scraper.scrapeForRegistration(registration, httpclient);

        Flux<Leveringslinje> nyeLandinger = landinger
                .filterWhen(leveringslinje -> forrigeLanding.map(leveringslinje.landingsdato::isAfter));

        // LocalDate forrigeLanding = repository.forrigeLandingFor(registration).block();
        // Flux<Leveringslinje> nyeLandinger = landinger.filter(l -> l.landingsdato.isAfter(forrigeLanding));
        logger.info("Nye landinger: {}", nyeLandinger);
        repository.lagreLeveranselinjer(nyeLandinger);
        logger.info("Done running scrapeForRegistrations for {}", registration);
    }

    void shutdown (){
        healthCheckRegistry.unregister("ScrapingJobHealthCheck");
        metricRegistry.remove("ScrapeTimerMetrics");
    }
}

File: plugins/scrape/src/main/java/no/lillehaug/landingsopplysninger/scrape/ScrapingJob.kt
package no.lillehaug.landingsopplysninger.scrape

import com.codahale.metrics.MetricRegistry
import com.codahale.metrics.Timer
import com.codahale.metrics.health.HealthCheck
import com.codahale.metrics.health.HealthCheckRegistry
import no.lillehaug.landingsopplysninger.api.LandingsopplysningerRepository
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.impl.client.HttpClients
import org.slf4j.LoggerFactory

class ScrapingJob(private val scraper: Scraper,
                  private val repository: LandingsopplysningerRepository,
                  val registrations: List<String>,
                  val healthCheckRegistry: HealthCheckRegistry,
                  val metricRegistry: MetricRegistry) {
    private val log = LoggerFactory.getLogger(javaClass)
    var previousRunSuccess = true
    private val timer = Timer()

    init {
        healthCheckRegistry.register("ScrapingJobHealthCheck", object: HealthCheck() {
            override fun check(): Result {
                return if(previousRunSuccess) {
                    Result.healthy()
                } else {
                    Result.unhealthy("ScrapingJob failed")
                }
            }
        })
        metricRegistry.register("ScrapeTimerMetrics", timer)
    }

    fun scrapeForRegistrations() {
        timer.time({
            try {
                val httpclient = HttpClients.createDefault();

                log.info("Running scrapeForRegistrations for {}", registrations)
                for (registration in registrations) {
                    scrapeForRegistration(registration, httpclient)
                }
                // TODO use trywr
                httpclient.close()
            } catch(e: Throwable) {
                previousRunSuccess = false
                log.error("Error scrapeForRegistrations", e)
            }
            log.info("Done running scrapeForRegistrations for {}", registrations)
        })
    }

    private fun scrapeForRegistration(registration: String, httpclient: CloseableHttpClient) {
        log.info("Running scrapeForRegistrations for {}", registration)
        val landinger = scraper.scrapeForRegistration(registration, httpclient)
        log.debug("Landinger: {}", landinger)
        val forrigeLanding = repository.forrigeLandingFor(registration)
        log.debug("Forrige landing: {}", forrigeLanding)
        val nyeLandinger = landinger.filter { it.landingsdato.isAfter(forrigeLanding) }
        log.info("Nye landinger: {}", nyeLandinger)
        repository.lagreLeveranselinjer(nyeLandinger)
        log.info("Done running scrapeForRegistrations for {}", registration)
    }

    fun shutdown (){
        healthCheckRegistry.unregister("ScrapingJobHealthCheck")
        metricRegistry.remove("ScrapeTimerMetrics")
    }
}
